### ISSUE TITLE
1.11.4 fix which returns default timezone if set for the alias 'local' breaks everything

### ISSUE THREAD

**Issue by cravindra** [moment/luxon#469]
Hello.

Until the recent update to `1.11.4` using `local` time zone as a string explicitly set the zone to the browser default. While this was unexpected initially, we worked around it as follows:

```js
// Set default time zone for project:
Settings.defaultZoneName = timezone || 'local';

// To interpret server side dates given in ISO:
const zonedDate = DateTime.fromISO(date); // Interprets it as the project timezone

// Show on the UI: Must be scoped to browser timezone so as to show it in terms of the local time
// which may be different from the project timezone (Ex: A user in Japan selecting shift timings in a 
// date input for a project in India - The user in Japan must be able to select times (say 8am to 3pm) 
// in his/her worldview without having to do conversions
const displayDate = zoneDate.setZone('local', {keepLocalTime: true});

// Parse input from the date-picker in browser zone, convert to project timezone, convert to ISO for server
const utc = DateTime
                                .fromJSDate(date, {zone: 'local'})
                                 // Get project time zone by checking the zone of .a new DateTime object
                                .setZone(DateTime.local().zoneName, {keepLocalTime: true})
                                .setZone('utc')
                                .toMillis()

```

Now obviously this has stopped working. I have to admit that the new behaviour should be the expected behaviour. I was initially surprised by the older behaviour and had to code around it. Now this fix is breaking all our code 😆 . While this is something which can be fixed on our end, I feel there is a legitimate need to be able to quick access the browser default.

Perhaps we can introduce a new alias like `local` such as `system`? Or is there a better solution I'm not seeing? (I'd like to avoid the use of the timeZoneOffset from the native date object although that is one feasible solution)

**Comment by icambron** [comment]
Yeah, I think that's a reasonable request. I hadn't even thought of this use case. I'd take a PR aliasing "system" to the local instance.

The workaround is to explicitly import the `LocalZone` object and pass `LocalZone.instance` to the `setZone` call.

 Also, just for completeness, I'll point out that you could also *not* set the default zone and just do:

 ```js
DateTime.fromISO(s, { zone: projectTimeZone }).toUTC()
```

**Comment by cravindra** [comment]
Raised a PR: https://github.com/moment/luxon/pull/475

Using `LocalZone.instance` is a good idea.. I was actually unaware of that class until I dug into the code.

Passing the timezone everywhere is not feasible as the entire codebase assumes once the default zone has been set it can be blindly assumed to be in that timezone. This means that the call may happen where that value is not explicitly available

**Comment by GillesDebunne** [comment]
I would argue that using `LocalZone.instance` is more explicit than introducing a new keyword. Overriding `defaultZone` is already a bit hacky, and this may feel like a patch on top of it.

However, some documentation is indeed missing around _local_ usage, with each time a link to the `defaultZone` paragraph (itself referencing all the methods impacted by this setting). With examples using `LocalZone.instance`, or resetting `Settings.defaultZoneName = null`, we can make it clear that `local` is indeed local, unless its global semantic is overridden, making the API feel right to me.

**Comment by icambron** [comment]
@GillesDebunne I merged the PR above, but I kind of agree with you. Thinking about it a little more, I think it's going to be pretty confusing. Actually what's confusing is that "local" refers to an overridable default, whereas LocalZone.instance does not. So what I should have done is had "default" and "local". I'm not quite show to extracate myself from that now.

**Comment by GillesDebunne** [comment]
I wonder if deprecating the "local" keyword altogether is an option.

Using `undefined` (i.e. no default) everywhere `local` is currently used, which would fall back to `LocalZone`, unless overridden in `Settings` like today.

When providing no specific value, it feels more natural to me to use a sensible default (local), unless explicitly overridden. Whereas specifically asking for a 'local' and sometimes ending up with a different value is surprising.

This could be a non breaking change before v2, just a change in the documentation. Reverting #475 before the next release would help.

**Comment by cravindra** [comment]
@GillesDebunne I think deprecating `local` is a good idea. `undefined` => what local does today (use the system default unless it has the default has been set using the Settings object)> This would preserve the existing behaviour as well and as you pointed out - only requires highlighting in the docs.

That being said, I still feel there is a case for `system` to explicitly set it to the system default. Without confusing the API by having the use to know of the inner classes like `LocalZone`. It causes an extra import to be required and a deep knowledge of the Luzon API. The alias `system` is cleaner and makes sense imo. (The argument of using `LocalZone.instance` can be made to remove the `utc` alias and use the `FixedTimeZoneOffset.getUtcInstance()` too but thats a little ungainly)

**Comment by GillesDebunne** [comment]
@cravindra We agree, and I now understand the limitations of using `LocalZone.instance`.

I checked how `local` is used in the code, and using `undefined` instead make sense to me.

I would introduce an other change: `toLocal()` would *always* set the zone to local (i.e system), independently of the `defaultZone`. 

The current implementation was consistent, where `local` meant "local except if overridden in Settings" in the constructors, and `toLocal()`, because of 'local' in its name, shall reset the _same_ value, i.e. depend on `defaultZone`.

However, with `undefined` as the default zone, `toLocal` no longer needs to do this, and can become much less surprising, with a consistent behavior independent of `defaultZone`.

Thinking more about it, I agree there is a case for `system`. However, I would call it `local` instead, for consistency with `toLocal()`, for familiarity with the current API and because I find it more explicit.

This leads me to this final proposal:
- `undefined` is used as the zone default value in all constructors and means "local unless `defaultZone` is set" (we agreed on this one).
- the string `local` can be used in these methods instead, and would explicitly always require a local zone.
- Consistently and symmetrically, `toLocal()` always sets the zone to local. We could add a new `toDefaultZone()` method for completeness.

What do you think?

**Comment by icambron** [comment]
> the string local can be used for the zone in these methods instead, and would explicitly always require a local zone.

The problem with this is local currently means default, so this would be breaking. I know I just broke things *to* that, but I think this would only compound that confusion. So what about this instead?

 1. deprecate `"local"` and `toLocal` and remove them from the docs
 2. add `"default"` and `"system"` strings
 3. add `toDefaultZone` and `toSystemZone`

**Comment by GillesDebunne** [comment]
I preferred `local` to `system`, but then realized I sometimes mock my test by changing my system settings. So `system` is maybe more accurate, since it might be different from local, but it is what is actually used.

My only concern then is the `local()` constructor. It feels weird and ambiguous to rename it to `system()`.

Also, I would prefer to avoid the `default` keyword, since it would be the same as `undefined`. I find the latter clearer: after all, that's what a default is for, i.e. used when no value is provided.

**Comment by icambron** [comment]
I don't think we should rename `local()` to `system()`. The point of `local()` is that it's produces a local time of some sort (actual zone tbd), not that it is a particular local time with the zone "local". So in many ways, removing "local" (the zone) makes things clearer for `local()` the method.

I don't like the `undefined` stuff much. `undefined` is fine to pass in and triggers the default arguments for things, but those defaults should still work:

```js
setZone(undefined)

// is really

setZone()

//same as

setZone("default")

//because it's

function setZone(zone = "default")
```

This "surfacing" of the options is really nice in a lot of situations where the arguments are being driven by some higher order logic in the applicatioon.

**Comment by GillesDebunne** [comment]
Upon reflection, I agree with this proposal and I'd be happy to contribute to the PR.

**Comment by GillesDebunne** [comment]
After a few months, I've started working on this PR.

The first part, using `default` instead of `local` in all DateTime constructors, and referring to _"system's time zone"_ instead of _"local zone"_ in the docs is done and feels like a step in the right direction.
After all the `default` zone refers to `Settings.defaultZone`, which is more obvious than `local`.

The next steps are to deprecate `toLocal()` and to replace it with `toDefaultZone` and `toSystemZone` (memo, Default equals to System unless overridden in Settings), and to introduce a new `system` keyword as a zone parameter.

`toDefaultZone()` is not ambiguous. However, there is a naming inconsistency with the `LocalZone` class: `.toSystemZone()` would actually be a shortcut to `.setZone(LocalZone.instance)`. 

Here are the options I can think of:
1. Use `toLocalZone()` instead of `toSystemZone()`. That's what I would prefer, but then the `system` zone keyword should then be named `local` for consistency, which is a breaking change.
2. Rename (or alias) `LocalZone` to `SystemZone`, and change its `type` and `name` methods.
3. Keep the name `toSystemZone()` even if it relates to `LocalZone`.
4. Postpone `toLocalZone()` or only the `local` keyword to v2.0.

While on it, since the question was raised, I'd argue that `local()` should rely on the *default* zone, for consistency with all the other constructors.

What are your thoughts on this?

**Comment by icambron** [comment]
@GillesDebunne Before you get too far, I took a stab at this in the 2.0 branch, which you can see here: https://github.com/moment/luxon/tree/2.0-take-2. I should have updated this ticket to reflect that; sorry if I wasted your time.

Anyway, in that branch, I actually just made the change of LocalZone -> SystemZone (as in `SystemZone.instance` to resolve that naming conflict: https://github.com/moment/luxon/blob/2.0-take-2/src/zones/systemZone.js

**Comment by GillesDebunne** [comment]
@icambron I had a look at this new 2.0 branch, and I really liked the removal of the invalid state and how it simplified the code.

Most of your changes align with what I started, and I'll rebase on your branch. Before that, I fixed a few things on your branch while reading your commits. PR soon to come.

What are your plans for 2.0? The documentation needs to be updated, but do you have other breaking changes in mind? What is the ETA?

**Comment by icambron** [comment]
@GillesDebunne The ETA is really when I get all the changes done, and I don't have a great sense of when I'll have time for that (day job super busy atm), and of course it depends if I get contributions. You can see the milestone here: https://github.com/moment/luxon/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.0

The biggest outstanding change is to split off the formatting and parsing stuff to its own package (probably all in the same repo, but separate NPMs). Once that's in I'll probably start pushing out prelease builds while making the other changes.

**Comment by icambron** [comment]
Fixed in 2.0
